### PR TITLE
[JENKINS-33950] dependencies may be installed multiple times

### PR DESCRIPTION
**Pull request by kzantow**
When installing multiple plugins (e.g. during the setup wizard, but also reproducible through the plugin manager) plugins may not be properly determined to be installed and may get installed multiple times, resulting in unnecssary restart required indicators. These installations happen in parallel based on downloaded plugin metadata, so it's not known until installation time whether another previously installed plugin ended up requiring the same thing. This fixes the main issue of Jenkins indicating it requires a restart, but doesn't prevent multiple concurrent downloads.

This fixes: https://issues.jenkins-ci.org/browse/JENKINS-33950

[Link for reviewers](2258/files?w=1)

@reviewbybees

### CODE REVIEW

**Comment by kzantow** [comment]
Note: I spent way too much time trying to come up with a unit test for this, the classes dealing with this are all final, inner... it would have required many more changes to the code than I thought were warranted for this fix.

**Comment by daniel-beck** [comment]
Showing the version on the UI seems unrelated. What's the issue for that?

**Comment by kzantow** [comment]
@daniel-beck it's mostly to validate that it only shows success when the version is identical

File: core/src/main/java/hudson/model/UpdateCenter.java [hunk 1]
```diff
@@ -1652,6 +1661,14 @@ public String getName() {
...
         @Override
         public void _run() throws IOException, InstallationStatus {
+            if (isAlreadyInstalling()) {
+                // Do this first so we can avoid duplicate downloads, too
+                // check to see if the plugin is already installed at the same version and skip it
+                LOGGER.warning("Skipping duplicate install of: " + plugin.getDisplayName() + "@" + plugin.version);
+                //throw new Skipped(); // TODO set skipped once we have a status indicator for it
+                return;
+            }
+            try {
```
Comment (kzantow) [inline core/src/main/java/hudson/model/UpdateCenter.java]: I'll fix the formatting before merge, this is to show the actual changes

File: core/src/main/java/hudson/model/UpdateCenter.java [hunk 2]
```diff
@@ -1676,6 +1693,50 @@ public void _run() throws IOException, InstallationStatus {
...
+                    notifyAll();
+                }
+            }
+        }
+
+        /**
+         * Indicates there is another installation job for this plugin
+         * @since TODO
+         */
+        protected boolean isAlreadyInstalling() {
```
Comment (kzantow) [inline core/src/main/java/hudson/model/UpdateCenter.java]: I should probably rename this method... something like `wasInstalled`, will do so before merging

Comment (recena) [inline core/src/main/java/hudson/model/UpdateCenter.java]: :+1: for renaming (maybe `isInstalled`)

File: core/src/main/java/hudson/model/UpdateCenter.java [hunk 3]
```diff
@@ -1656,29 +1665,81 @@ public String getName() {
 
         @Override
         public void _run() throws IOException, InstallationStatus {
-            super._run();
+            if (wasInstalled()) {
+                // Do this first so we can avoid duplicate downloads, too
+                // check to see if the plugin is already installed at the same version and skip it
+                LOGGER.warning("Skipping duplicate install of: " + plugin.getDisplayName() + "@" + plugin.version);
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/UpdateCenter.java]: Warning seems a bit much here, as this appears to be a fairly common case. INFO, or probably even lower, seems more appropriate.

**Pull request merged**